### PR TITLE
feat: view photos on reviews

### PR DESCRIPTION
**Pull request by anormananderson**
Currently the signed urls associated with each image are displayed inside each post, but no functionality exists to display the actual images.

Also, for an unknown reason, sometimes no urls are returned.

Branch off of #111

### CODE REVIEW

**Comment by anormananderson** [comment]
@callumcurtis

File: apps/nextjs/next.config.js [hunk 1]
```diff
@@ -15,6 +15,17 @@ const config = {
     "@restauwants/validators",
   ],
 
+  images: {
+    remotePatterns: [
+      {
+        protocol: "https",
+        hostname: "storage.googleapis.com",
+        pathname: "/restauwants_production/*",
```
Comment (callumcurtis) [inline apps/nextjs/next.config.js:23]: Can "restauwants_production" (the bucket name) be an env variable instead? it's used in a bunch of other places as well.

File: apps/nextjs/src/app/api/photo/route.ts [hunk 2]
```diff
@@ -0,0 +1,291 @@
+import type { NextRequest } from "next/server";
+import { Storage } from "@google-cloud/storage";
+import { v4 as uuidv4 } from "uuid";
+import { z } from "zod";
+
+const MAX_FILE_SIZE = 5 * 1024 * 1024; // 5 MB
```
Comment (callumcurtis) [inline apps/nextjs/src/app/api/photo/route.ts:6]: this is hardcoded elsewhere in a validator - it should probably be put in a central location and reused

File: apps/nextjs/src/app/api/photo/route.ts [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+await setCors();
+
+async function setCors() {
+  // Set the cors config once
+  const storage = await getStorage();
+  const stagingBucket = storage.bucket("restauwants_staging");
```
Comment (callumcurtis) [inline apps/nextjs/src/app/api/photo/route.ts:12]: these bucket names should probably be env vars eventually

File: apps/nextjs/src/app/api/photo/route.ts [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+
+  await stagingBucket.setCorsConfiguration([config]);
+  console.log("Set cors config for staging bucket.");
+  await prodBucket.setCorsConfiguration([config]);
+  console.log("Set cors config for production bucket.");
+}
+
+async function getStorage() {
+  return new Storage({
+    projectId: "nifty-edge-415820",
```
Comment (callumcurtis) [inline apps/nextjs/src/app/api/photo/route.ts:30]: seems like an env var

File: apps/nextjs/src/app/api/photo/route.ts [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+  const size = searchParams.get("size");
+  if (!signedUrl || !size) {
+    return Response.json(
+      { error: "signedUrl and size query parameters are required" },
+      { status: 400 },
+    );
+  }
+
+  const signedUrlString = z.string().parse(signedUrl);
+  const sizeInt = z
```
Comment (callumcurtis) [inline apps/nextjs/src/app/api/photo/route.ts:68]: could make this (or a part of this) a validator in the validators package and reuse it elsewhere

File: apps/nextjs/src/app/api/photo/route.ts [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+    .string()
+    .pipe(z.coerce.number().nonnegative().lte(MAX_FILE_SIZE))
+    .parse(size);
+
+  moveFileIfVerified(
+    signedUrlString,
+    sizeInt,
+    "restauwants_staging",
```
Comment (callumcurtis) [inline apps/nextjs/src/app/api/photo/route.ts:76]: probably should be env vars

File: apps/nextjs/src/app/api/photo/route.ts [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+  if (!filename) {
+    return Response.json(
+      { error: "filename query parameter is required" },
+      { status: 400 },
+    );
+  }
+
+  filename = z.string().uuid().parse(filename);
+
+  deleteFileIfAuthenticated(filename, "restauwants_production").catch(() =>
```
Comment (callumcurtis) [inline apps/nextjs/src/app/api/photo/route.ts:96]: probably should be an env var. I'll stop mentioning these though.

File: apps/nextjs/src/app/api/photo/route.ts [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+
+async function upload(searchParams: URLSearchParams) {
+  const size = searchParams.get("size");
+  if (!size) {
+    return Response.json(
+      { error: "Size query parameter is required" },
+      { status: 400 },
+    );
+  }
+  const sizeInt = z
```
Comment (callumcurtis) [inline apps/nextjs/src/app/api/photo/route.ts:124]: seems like this validator could be defined once and reused above and here

File: apps/nextjs/src/app/api/photo/route.ts [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+  // check if the file exists first
+  const file = bucket.file(filename);
+  if (!(await file.exists())) {
+    throw new Error("File to be moved was not uploaded to bucket.");
+  } else {
+    await file.delete();
+  }
+}
+
+function getBaseUrl() {
```
Comment (callumcurtis) [inline apps/nextjs/src/app/api/photo/route.ts:288]: this looks copy and paste from the trpc file - should be extracted and shared

File: apps/nextjs/src/components/reviews.tsx [hunk 10]
```diff
@@ -69,6 +71,56 @@ export function CreateReviewForm() {
 
   const utils = api.useUtils();
 
+  const createPhotos = api.photo.addPhotosToReview.useMutation({
```
Comment (callumcurtis) [inline apps/nextjs/src/components/reviews.tsx:74]: not clear what the difference between "create" and "upload" photos is

File: apps/nextjs/src/components/reviews.tsx [hunk 11]
```diff
@@ -79,8 +131,33 @@ export function CreateReviewForm() {
...
+    const sizes: number[] = [];
+    // Upload every file from the form
+    for (const file of files) {
+      try {
+        const res = await uploadFile(file);
+        urls.push(res.url);
+        sizes.push(file.size);
+      } catch (error) {
+        toast.error("Failed to upload photos");
+        console.error(error);
```
Comment (callumcurtis) [inline apps/nextjs/src/components/reviews.tsx:146]: remove

File: apps/nextjs/src/components/reviews.tsx [hunk 12]
```diff
@@ -79,8 +131,33 @@ export function CreateReviewForm() {
...
+        return;
```
Comment (callumcurtis) [inline apps/nextjs/src/components/reviews.tsx:147]: does returning here work with the "submit" handling? do you need to throw an error instead?

File: apps/nextjs/src/components/reviews.tsx [hunk 13]
```diff
@@ -79,8 +131,33 @@ export function CreateReviewForm() {
...
+
+    // add the review
+    try {
+      const reviewId = await createReview.mutateAsync(
+        CreateReviewSchema.parse(data),
+      );
+      await addPhotosToReview(urls, sizes, reviewId);
+    } catch (error) {
+      toast.error("Failed to add photos to review");
+      console.error(error);
```
Comment (callumcurtis) [inline apps/nextjs/src/components/reviews.tsx:159]: remove

File: apps/nextjs/src/components/reviews.tsx [hunk 14]
```diff
@@ -396,6 +541,21 @@ export function ReviewCard(props: {
...
+              <div
+                className="relative aspect-square object-cover sm:size-full xl:w-6/12"
+                key={photo}
+              >
+                <Image
+                  src={photo}
+                  alt="Review photo"
+                  fill
+                  style={{
+                    objectFit: "cover", // cover, contain, none
```
Comment (callumcurtis) [inline apps/nextjs/src/components/reviews.tsx:554]:
```suggestion
                    objectFit: "cover",
```

File: packages/api/src/router/photo.ts [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+
+import { createTRPCRouter, protectedProcedure } from "../trpc";
+
+export const photoRouter = createTRPCRouter({
+  getSignedUploadUrl: protectedProcedure
+    .input(
+      z
+        .number()
+        .nonnegative()
+        .lt(3 * 1024 * 1024),
```
Comment (callumcurtis) [inline packages/api/src/router/photo.ts:16]: should make a common validator

File: packages/api/src/router/photo.ts [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+    )
+    .mutation(async ({ input }) => {
+      const res = await fetch(
+        getBaseUrl() + `/api/photo?action=upload&size=${input}`,
+      );
+      const { presignedUrl } = (await res.json()) as { presignedUrl: string };
+      return presignedUrl;
+    }),
+
+  /*
```
Comment (callumcurtis) [inline packages/api/src/router/photo.ts:26]: remove dead code

File: packages/api/src/router/photo.ts [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+      } else if (review.userId !== ctx.session.user.id) {
+        throw new Error("Not authorized");
+      }
+      await ctx.db.delete(schema.photo).where(eq(schema.photo.id, input));
+    }),
+});
+
+// PRECONDITION: The user must own the review
+async function addPhotosToReviewInDB(
+  ctx: {
```
Comment (callumcurtis) [inline packages/api/src/router/photo.ts:111]: just pass the db. See the friend router for examples.

File: packages/api/src/router/photo.ts [hunk 18]
```diff
@@ -0,0 +1,185 @@
...
+  },
+  reviewId: number,
+  photoIds: string[],
+) {
+  interface row {
+    reviewId: number;
+    id: string;
+    createdAt: Date;
+  }
+  const rows = [] as row[];
```
Comment (callumcurtis) [inline packages/api/src/router/photo.ts:128]:
```suggestion
  const rows: row[] = [];
```
don't use "as", prefer colon (should search your code for other uses)

File: packages/api/src/router/photo.ts [hunk 19]
```diff
@@ -0,0 +1,185 @@
...
+        id: string;
+      };
+      expires: string;
+    };
+    db: Database;
+  },
+  reviewId: number,
+  photoIds: string[],
+) {
+  interface row {
```
Comment (callumcurtis) [inline packages/api/src/router/photo.ts:123]: you should create a Zod validator in packages/validators/src/db/index.ts instead

File: packages/api/src/router/photo.ts [hunk 20]
```diff
@@ -0,0 +1,185 @@
...
+    if (!res.ok) {
+      throw new Error("Failed to fetch presigned URL");
+    }
+    const { presignedUrl } = (await res.json()) as { presignedUrl: string };
+    return presignedUrl;
+  });
+  return Promise.all(signedUrls);
+};
+
+function getBaseUrl() {
```
Comment (callumcurtis) [inline packages/api/src/router/photo.ts:181]: this always runs on the server, no? so the window will never be defined

File: packages/api/src/router/review.ts [hunk 21]
```diff
@@ -26,22 +30,35 @@ export const reviewRouter = createTRPCRouter({
...
+            )
+            .where(where)
+            .orderBy(desc(schema.review.id))
+            .limit(input.limit + 1)
+        ).map(async (r) => ({
+          ...r.review,
+          photos: await photoUrlsFromReviewId(
+            ctx.db,
+            r.review.id,
+            // TODO: Think about handling this better rather than failing silently
```
Comment (callumcurtis) [inline packages/api/src/router/review.ts:54]: I think you should filter out any reviews that fail like this

File: packages/api/src/router/review.ts [hunk 22]
```diff
@@ -70,21 +87,31 @@ export const reviewRouter = createTRPCRouter({
...
```
Comment (callumcurtis) [inline packages/api/src/router/review.ts:107]: same as above - filter out reviews that fail like this

File: packages/api/src/router/review.ts [hunk 23]
```diff
@@ -95,14 +122,17 @@ export const reviewRouter = createTRPCRouter({
   create: protectedProcedure
     .input(CreateReviewSchemaExternal)
     .mutation(async ({ ctx, input }) => {
-      await ctx.db.insert(schema.review).values(
+      // create the review
+      const reviewTable = await ctx.db.insert(schema.review).values(
```
Comment (callumcurtis) [inline packages/api/src/router/review.ts:126]:
```suggestion
      const review = await ctx.db.insert(schema.review).values(
```
the result isn't the whole table

File: packages/db/src/schema/index.ts [hunk 24]
```diff
@@ -198,3 +198,16 @@ export const friendRequestRelations = relations(friendRequest, ({ one }) => ({
...
+export const photo = mySqlTable("photo", {
+  reviewId: int("reviewId").notNull(),
+  id: varchar("id", { length: 255 }).notNull().primaryKey(),
+  createdAt: timestamp("createdAt")
+    .notNull()
+    .default(sql`CURRENT_TIMESTAMP`),
+});
+
+export const photoRelations = relations(photo, ({ one }) => ({
+  user: one(users),
```
Comment (callumcurtis) [inline packages/db/src/schema/index.ts:211]: doesn't link directly to user

File: packages/db/src/schema/index.ts [hunk 25]
```diff
@@ -198,3 +198,16 @@ export const friendRequestRelations = relations(friendRequest, ({ one }) => ({
...
+  review: one(review),
```
Comment (callumcurtis) [inline packages/db/src/schema/index.ts:212]: missing relation in the opposite direction - from review to photos

File: packages/validators/src/client/index.ts [hunk 26]
```diff
@@ -26,6 +27,23 @@ export const CreateReviewFormSchema = z.object({
...
+  files: z
+    .custom<Blob>((value) => value instanceof Blob, {
+      message: "File must be a blob",
+    })
+    .array()
+    .max(7)
+    .refine(
+      (files) => {
+        for (const file of files) {
+          if (file.size === 0 || file.size > 5 * 1024 * 1024) {
```
Comment (callumcurtis) [inline packages/validators/src/client/index.ts:39]: file size should be a shared constant

File: packages/validators/src/server/external/index.ts [hunk 27]
```diff
@@ -66,3 +66,14 @@ export const RevokeFriendRequestSchema = z.object({
 export const CreateProfileSchema = z.object({
   username: username,
 });
+
+export const CreatePhotosSchema = z.object({
+  reviewId: z.number().int().positive(),
+  signedUrls: z.string().array(),
```
Comment (callumcurtis) [inline packages/validators/src/server/external/index.ts:72]: could define the allowed length

File: packages/validators/src/server/external/index.ts [hunk 28]
```diff
@@ -66,3 +66,14 @@ export const RevokeFriendRequestSchema = z.object({
...
+  sizes: z
+    .number()
+    .int()
+    .positive()
+    .lte(3 * 1024 * 1024)
```
Comment (callumcurtis) [inline packages/validators/src/server/external/index.ts:77]: should be a shared sizes constant

File: packages/validators/src/server/index.ts
Comment (callumcurtis) [inline packages/validators/src/server/index.ts:1]: delete this empty file

**Review by callumcurtis — CHANGES_REQUESTED** [review]

**Comment by anormananderson** [comment]
see #139

**Pull request closed without merging**